core/routing_spray.go

```diff
@@ -0,0 +1,313 @@
+		}).Debug("Not relaying bundle because there are copies left")
+		return nil, false
+	}
+
+	for _, cs := range sw.c.convergenceSenders {
+		// if we ran out of copies, then don't send it to any further peers
+		if metadata.remainingCopies < 2 {
+			break
+		}
+
+		var skip = false
+		for _, eid := range metadata.sent {
+			if cs.GetPeerEndpointID() == eid {
+				skip = true
+				break
+			}
+		}
+
+		if !skip {
+			css = append(css, cs)
+			metadata.sent = append(metadata.sent, cs.GetPeerEndpointID())
+			metadata.remainingCopies = metadata.remainingCopies - 1
+		}
+	}
+
+	sw.bundleData[bp.ID()] = metadata
+
+	log.WithFields(log.Fields{
+		"bundle":              bp.ID(),
+		"convergence-senders": css,
+		"remaining copies":    metadata.remainingCopies,
+	}).Debug("SprayAndWait selected Convergence Senders for an outgoing bundle")
+
+	del = false
+	return
+}
+
+// ReportFailure re-increments remaining copies if delivery was unsuccessful.
+func (sw SprayAndWait) ReportFailure(bp BundlePack, sender cla.ConvergenceSender) {
+	log.WithFields(log.Fields{
+		"bundle":  bp.ID(),
+		"bad_cla": sender,
+	}).Debug("Transmission failure")
+
+	metadata, _ := sw.bundleData[bp.ID()]
+	metadata.remainingCopies = metadata.remainingCopies + 1
+
+	for i := 0; i < len(metadata.sent); i++ {
+		if metadata.sent[i] == sender.GetPeerEndpointID() {
+			metadata.sent = append(metadata.sent[:i], metadata.sent[i+1:]...)
+			break
+		}
+	}
+
+	sw.bundleData[bp.ID()] = metadata
+}
+
+// BinarySpray implements the binary Spray and Wait routing protocol
+// In this case, each node hands over floor(copies/2) during the spray phase
+type BinarySpray struct {
+	c *Core
+	// bundleData stores the metadata for each bundle
+	bundleData map[string]sprayMetaData
+}
+
+// NewBinarySpray creates new instance of BinarySpray
+func NewBinarySpray(c *Core) BinarySpray {
+	log.WithFields(log.Fields{
+		"L": sprayl,
+	}).Debug("Initialised BinarySpray")
+	// register our custom metadata-block
+	extensionBlockManager := bundle.GetExtensionBlockManager()
+	if !extensionBlockManager.IsKnown(ExtBlockTypeBinarySprayBlock) {
+		// since we already checked if the block type exists, this really shouldn't ever fail...
+		_ = extensionBlockManager.Register(NewBinarySprayBlock(0))
+	}
+	return BinarySpray{
+		c:          c,
+		bundleData: make(map[string]sprayMetaData),
+	}
+}
+
+// NotifyIncoming tells the routing algorithm about new bundles.
+// In this case, we check, whether we are the originator of this bundle
+// If yes, then we initialise the remaining Copies to L
+// If not we attempt to ready the routing-metadata-block end get the remaining copies
+func (bs BinarySpray) NotifyIncoming(bp BundlePack) {
+	if metadataBlock, err := bp.Bundle.ExtensionBlock(ExtBlockTypeBinarySprayBlock); err == nil {
+		binarySprayBlock := metadataBlock.Value.(*BinarySprayBlock)
+		metadata := sprayMetaData{
+			sent:            make([]bundle.EndpointID, 0),
+			remainingCopies: binarySprayBlock.RemainingCopies(),
```

Comment:
At this point a small sanity check should be carried out, e.g., compare `RemainingCopies` against our `L`.

Reply:
Wöntfix